tor_archivist/core/blossom.py

```diff
@@ -0,0 +1,74 @@
+import requests
+
+class ConfigurationError(Exception):
+    pass
+
+class BlossomAPI(object):
+    def __init__(
+            self,
+            email: str=None,
+            password: str=None,
+            api_key: str=None,
+            api_base_url: str="http://api.grafeas.localhost:8000",
+            login_url: str="http://grafeas.localhost:8000/login/"
+    ):
+        if not email:
+            raise ConfigurationError("Need an email address!")
+        if not password:
+            raise ConfigurationError("Need a password!")
+        if not api_key:
+            raise ConfigurationError("Need an API key!")
+
+        self.email = email
+        self.password = password
+        self.base_url = api_base_url
+        self.login_url = login_url
+
+        self.s = requests.Session()
+        self.s.headers.update({'Authorization': f'Api-Key {api_key}'})
+
+    def _login(self):
+        resp = self.s.post(
+            self.login_url, data={
+                'email': self.email, 'password': self.password
+            }
+        )
+        return resp
+
+    def _call(self, method: str, path: str, data=None, json=None, params=None):
+        # https://2.python-requests.org/en/master/user/advanced/#prepared-requests
+        req = requests.Request(
+            method=method,
+            url=self.base_url+path,
+            json=json,
+            data=data,
+            params=params
+        )
+
+        for _ in range(3):
+            prepped = self.s.prepare_request(req)
+            settings = self.s.merge_environment_settings(prepped.url, {}, None, None, None)
+            resp = self.s.send(prepped, **settings)
+
+            if resp.status_code == 403:
+                if resp.json().get('detail') == "Authentication credentials were not provided.":
+                    self._login()
```

Comment:
ooooh! Prepared requests... 🤤

I really like this approach you've taken with retrying to login if given a 403 error. Why are we retrying the `self._login()` part multiple times, as the `range(3)` loop thing implies?

I can't imagine logging in for a second or third time actually helping anything with a 403 response.

Reply:
@itsthejoker thoughts?

Reply:
I went with three attempts for this reason:

Attempt one: will probably need to log in. Do so.
Attempt two: perhaps something went wrong or had network issues.
Attempt three: last try to do the thing we originally tried to do in attempt two.

Then give up and raise issues.

Reply:
It's a minor nit, but my thought is this: Let's say there's some part of the API that's off-limits to the bot and we accidentally try to call it. Doing so would look like this:

1. call endpoint
2. failure response with 403 error code
3. send login creds to login endpoint
4. get logged-in response (set cookies, etc.)
5. call endpoint (second time)
6. failure response with 403 error code
7. send login creds to login endpoint
8. get logged-in response (set cookies, etc.)
9. call endpoint (third and final time)
10. failure response with 403 error code
11. send login creds to login endpoint
12. get logged-in response (set cookies, etc.)
13. end of loop, overall failure of all calls

This means we'd be retrying to login 2 more times than necessary, and as a reaction that happens regardless of if there's another iteration of the loop instead of occurring just-in-time prior to the event requiring authorization.

---

Really this comes around to the idea that logging in as a reaction to 403 response sounds super handy from a developer experience standpoint, but that 403 response is useful and there are a _lot_ of edge cases. Might be better to offload handling the 403 response outside of the client library.